iree/tools/utils/image_util.c

```diff
@@ -120,57 +123,72 @@ iree_status_t iree_tools_utils_buffer_view_from_image(
-          element_byte * buffer_length, &mapped_memory);
-      if (!iree_status_is_ok(result)) {
-        iree_hal_buffer_release(buffer);
-        stbi_image_free((void*)pixel_data);
-        return result;
-      }
-      result = iree_tools_utils_pixel_rescaled_to_buffer(
-          pixel_data, buffer_length, (float*)mapped_memory.contents.data);
-      if (!iree_status_is_ok(result)) {
-        iree_hal_buffer_release(buffer);
-        stbi_image_free((void*)pixel_data);
-        return result;
-      }
-      iree_hal_buffer_unmap_range(&mapped_memory);
-      result = iree_hal_buffer_view_create(buffer, shape, shape_rank,
-                                           element_type, out_buffer_view);
-      iree_hal_buffer_release(buffer);
-      break;
-    }
-    default: {
-      stbi_image_free((void*)pixel_data);
-      return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
-                              "element type not supported");
-    }
+  // SINT_8 and UINT_8 perform direct buffer wrap.
+  result = iree_hal_buffer_view_wrap_or_clone_heap_buffer(
+      allocator, shape, shape_rank, element_type,
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      IREE_HAL_MEMORY_ACCESS_READ, IREE_HAL_BUFFER_USAGE_ALL,
+      iree_make_byte_span((void*)pixel_data, element_byte * buffer_length),
+      iree_allocator_null(), out_buffer_view);
+
+  stbi_image_free(pixel_data);
+  return result;
+}
+
+iree_status_t iree_tools_utils_buffer_view_from_image_rescaled(
+    const iree_string_view_t filename, const iree_hal_dim_t* shape,
+    iree_host_size_t shape_rank, iree_hal_element_type_t element_type,
+    iree_hal_allocator_t* allocator, const float* input_range,
+    iree_host_size_t range_length, iree_hal_buffer_view_t** out_buffer_view) {
+  *out_buffer_view = NULL;
+
+  if (element_type != IREE_HAL_ELEMENT_TYPE_FLOAT_32) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "element type should be f32");
   }
+
+  iree_status_t result;
+  uint8_t* pixel_data = NULL;
+  iree_host_size_t buffer_length;
+  result = iree_tools_utils_load_pixel_data(
+      filename, shape, shape_rank, element_type, &pixel_data, &buffer_length);
+  if (!iree_status_is_ok(result)) {
+    return result;
+  }
+
+  iree_host_size_t element_byte = iree_hal_element_byte_count(element_type);
+  iree_hal_buffer_t* buffer = NULL;
+  result = iree_hal_allocator_allocate_buffer(
+      allocator,
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+      IREE_HAL_BUFFER_USAGE_ALL, element_byte * buffer_length, &buffer);
+  if (!iree_status_is_ok(result)) {
+    stbi_image_free((void*)pixel_data);
+    return result;
+  }
+  // Need to normalize to the expected input range.
+  iree_hal_buffer_mapping_t mapped_memory;
+  result =
+      iree_hal_buffer_map_range(buffer, IREE_HAL_MEMORY_ACCESS_DISCARD_WRITE, 0,
+                                element_byte * buffer_length, &mapped_memory);
```

Comment:
(continuing the comment thread)
I'm not sure about the memory access here either. @benvanik ?

Reply:
what aren't you sure about? _pretend_ I have not read the comment thread.

Reply:
When we map the hal buffer here the content would be modified with the rescaled pixel data, so I think mark the access type to DISCARD_WRITE signals the content would be overridden. However, after the rescaling, the buffer will be used as the input data, and from [the runtime example](https://github.com/google/iree/blob/25f8064dcc9e3faa7f582db4b5b99e346d3866c1/iree/runtime/demo/hello_world_explained.c#L210), it should be marked as READ to indicate it is read only, similar to the above `iree_tools_utils_buffer_view_from_image`. Should we update the access type here?

Reply:
ah - the access here is just for this particular map_range - so in the future when the buffer is mapped again it can be mapped with different access modes - this is fine as-is!